src/main/java/uk/nhs/hee/trainee/details/mapper/CctMapper.java

```diff
@@ -82,4 +75,15 @@ public interface CctMapper {
    */
   @Mapping(target = "traineeId", source = "traineeId")
   CctCalculation toEntity(CctCalculationDetailDto dto, String traineeId);
+
+  /**
+   * Calculate the CCT end date for a CCT Calculation.
+   *
+   * @param entity The CctCalculation to use to calculate the CCT end date.
+   * @return the CCT end date, or null if CctCalculation is null.
+   */
+  @Named("calculateCctDate")
+  static LocalDate calculateCctDate(CctCalculation entity) {
+    return PLACEHOLDER_CCT_DATE;
+  }
```

Comment:
This is fine for now, but may not work well when we start doing real/complex calculations.

It might get a bit heavy to squish in to a mapper and calling a service from the mapper usually leads us to circular initialization issues at some point. So we may still have to pass the pre-calculated value as a parameter when mapping to the DTO,

Reply:
Yeah, fair enough. This is the current look of the real calculation, would you rather this lived outside of the mapper?

```
@Named("calculateCctDate")
  static LocalDate calculateCctDate(CctCalculation entity) {
    LocalDate currentEndDate = entity.programmeMembership().endDate();
    //assumption: changes are in sequential date order
    for (int i = 0; i < entity.changes().size(); i++) {
      CctCalculation.CctChange c = entity.changes().get(i);
      LocalDate startDate = c.startDate();
      if (startDate.isAfter(currentEndDate)) {
        Logger logger = LoggerFactory.getLogger(CctMapper.class);
        logger.warn("CCT date cannot be calculated, start date for change {} after end date {}.",
            c, currentEndDate);
        return null;
      }
      int chunkDays = Period.between(startDate, currentEndDate).getDays();
      double currentWte = i == 0
          ? entity.programmeMembership().wte()
          : entity.changes().get(i - 1).wte();
      double wte = c.wte();
      int chunkDaysWte = (int) Math.ceil((chunkDays * currentWte) / wte);
      currentEndDate = currentEndDate.plusDays(chunkDaysWte);
    }
    return currentEndDate;
  }
```

Reply:
Before we started using records we'd have just done something like

```java
LocalDate cctDate = calculateCctDate(entity);
CctCalculationDto dto = mapper.toDto(entity);
dto.setCctDate(cctDate)
```

rather than having the mapper do it, so it does feel like

```java
LocalDate cctDate = calculateCctDate(entity);
CctCalculationDto dto = mapper.toDto(entity, cctDate);
```

is the right way to do it in a record-based world